### PR TITLE
Launch/no build case

### PR DESCRIPTION
**Pull request by KyleGoyette**
Fixes WB-NNNN
Fixes #NNNN

Description
-----------
What does the PR do?

Testing
-------
How was this PR tested?

Checklist
-------
- Name PR "[WB-NNNN][WB-MMMM] Add support for..." similar to entries in CHANGELOG.md
- Include reference to internal ticket "Fixes WB-NNNN" (and github issue "Fixes #NNNN" if applicable)

### CODE REVIEW

File: wandb/sdk/launch/_project_spec.py [hunk 1]
```diff
@@ -427,11 +435,13 @@ def fetch_and_validate_project(
...
 
     """
+    if launch_project.source == LaunchSource.DOCKER:
+        return launch_project
     if launch_project.source == LaunchSource.LOCAL:
         if not launch_project._entry_points:
             wandb.termlog("Entry point for repo not specified, defaulting to main.py")
             launch_project.add_entry_point("main.py")
-    else:
+    elif launch_project.source != LaunchSource.DOCKER:
```
Comment (stephchen) [inline wandb/sdk/launch/_project_spec.py:444]: this `elif` is redundant given the return above

File: wandb/sdk/launch/launch.py [hunk 2]
```diff
@@ -71,7 +71,9 @@ def _run(
     runner_config: Dict[str, Any] = {}
     runner_config[PROJECT_SYNCHRONOUS] = synchronous
     runner_config[PROJECT_DOCKER_ARGS] = (
-        launch_config["docker"] if launch_config else {}
+        launch_config.get(launch_config.get("docker", {}).get("args", None), {})
```
Comment (stephchen) [inline wandb/sdk/launch/launch.py:74]: this line is super confusing can we break this up lol

File: wandb/sdk/launch/runner/kubernetes.py [hunk 3]
```diff
@@ -239,7 +233,8 @@ def run(self, launch_project: LaunchProject) -> Optional[AbstractRun]:  # noqa:
             if registry:
                 repo, tag = image_uri.split(":")
                 docker.push(repo, tag)
-
+        env_vars = get_env_vars_dict(launch_project, self._api)
+        containers[0]["env"] = [{"name": k, "value": v} for k, v in env_vars.items()]
```
Comment (stephchen) [inline wandb/sdk/launch/runner/kubernetes.py:237]: we should also have a line here allowing users to specify additional env vars

File: wandb/sdk/launch/runner/local.py [hunk 4]
```diff
@@ -75,23 +77,18 @@ def run(self, launch_project: LaunchProject) -> Optional[AbstractRun]:
...
             _logger.info("Pulling user provided docker image")
-            pull_docker_image(launch_project.docker_image)
-
-            wandb.termwarn(
-                "Using supplied docker image: {}. Artifact swapping and launch metadata disabled".format(
-                    launch_project.docker_image
-                )
-            )
+            if not docker_image_exists(launch_project.docker_image):
+                pull_docker_image(launch_project.docker_image)
```
Comment (stephchen) [inline wandb/sdk/launch/runner/local.py:85]: nice catch

File: wandb/sdk/launch/runner/local.py [hunk 5]
```diff
@@ -75,23 +77,18 @@ def run(self, launch_project: LaunchProject) -> Optional[AbstractRun]:
...
             image_uri = launch_project.docker_image
 
         else:
             # build our own image
             image_uri = construct_local_image_uri(launch_project)
+            entry_point = launch_project.get_single_entry_point()
```
Comment (stephchen) [inline wandb/sdk/launch/runner/local.py:91]: is this used?

File: wandb/sdk/launch/templates/_wandb_bootstrap.py [hunk 6]
```diff
@@ -64,8 +64,10 @@ def main():
             for req in f:
                 if len(ONLY_INCLUDE) == 0 or req.split("=")[0].lower() in ONLY_INCLUDE:
                     # can't pip install wandb==0.*.*.dev1 through pip. Lets just install wandb for now
-                    if req.startswith("wandb==") and "dev1" in req:
-                        req = "wandb"
+                    # if req.startswith("wandb==") and "dev1" in req:
+                    #     req = "wandb"
+                    if "wandb" in req:
+                        req = "git+https://github.com/wandb/client.git@launch/no-build-case"
```
Comment (stephchen) [inline wandb/sdk/launch/templates/_wandb_bootstrap.py:70]: add a note here to fix this in master

Comment (KyleGoyette) [reply wandb/sdk/launch/templates/_wandb_bootstrap.py:70]: Good call

File: wandb/sdk/launch/launch.py (hunk 2 above)
Comment (KyleGoyette) [reply wandb/sdk/launch/launch.py:74]: good call

File: wandb/sdk/launch/runner/kubernetes.py (hunk 3 above)
Comment (KyleGoyette) [reply wandb/sdk/launch/runner/kubernetes.py:237]: Sounds good

File: wandb/sdk/launch/runner/local.py (hunk 4 above)
Comment (KyleGoyette) [reply wandb/sdk/launch/runner/local.py:85]: ah, this isn't necessary anymore, we do the check in pull_docker_image

File: wandb/sdk/launch/runner/local.py (hunk 5 above)
Comment (KyleGoyette) [reply wandb/sdk/launch/runner/local.py:91]: Yep, in the call to generate_docker_image below this line

**Review by stephchen — APPROVED** [review]

**Pull request merged**